File: dart/lib/src/v1/utils/number_validator.dart
class NumberValidator {
  static final RegExp _regex = RegExp(r'^\+?(0|\d{3})\d{9}$');

  static List<String> validateNumbers(List<String> numbers) {
    if (numbers.isEmpty) {
      print('Number list cannot be null or empty');
      return [];
    }

    final cleansed = <String>{};
    for (var number in numbers) {
      if (number.trim().isEmpty) {
        print('Number ($number) cannot be null or empty!');
        continue;
      }
      number = number.trim().replaceAll(RegExp(r'-|\s'), '');
      if (_regex.hasMatch(number)) {
        if (number.startsWith('0')) {
          number = '256${number.substring(1)}';
        } else if (number.startsWith('+')) {
          number = number.substring(1);
        }
        cleansed.add(number);
      } else {
        print('Number ($number) is not valid!');
      }
    }
    return cleansed.toList();
  }
}

File: kotlin/src/main/kotlin/v1/utils/NumberValidator.kt
package v1.utils

object NumberValidator {
    private const val regex = "^\\+?(0|\\d{3})\\d{9}$"

    /**
     * Validates a list of phone numbers.
     *
     *
     * Handles conversion of numbers starting with '0' to '256' and removes duplicates.
     * Removes leading '+' signs and trims whitespace.
     *
     * @param numbers List of number inputs to validate.
     * @return A clean list of numbers, with duplicates removed and formatted correctly.
     */
    fun validateNumbers(numbers: List<String>): List<String> {
        if (numbers.isEmpty()) {
            System.err.println("Number list cannot be empty")
            return ArrayList()
        }

        val cleansed: MutableSet<String> = HashSet()
        for (number in numbers) {
            var number = number
            if (number.trim { it <= ' ' }.isEmpty()) {
                System.out.printf("Number (%s) cannot be empty!\n", number)
                continue
            }
            number = number.trim { it <= ' ' }.replace("-|\\s".toRegex(), "")
            if (number.matches(regex.toRegex())) {
                if (number.startsWith("0")) {
                    number = "256" + number.substring(1)
                } else if (number.startsWith("+")) {
                    number = number.substring(1)
                }
                cleansed.add(number)
            } else {
                System.out.printf("Number (%s) is not valid!\n", number)
            }
        }
        return ArrayList(cleansed)
    }
}
